File: 2019/php/src/Aoc/Days/Day7.php
<?php

namespace Ppx17\Aoc2019\Aoc\Days;

use Illuminate\Support\Collection;
use Ppx17\Aoc2019\Aoc\Days\Day7\IntCode;

class Day7 extends AbstractDay
{
    private IntCode $intCode;
    private Collection $combinations;

    public function dayNumber(): int
    {
        return 7;
    }

    public function setUp(): void
    {
        $this->intCode = new IntCode(array_map('intval', explode(',', $this->getInput())));
        $this->combinations = collect($this->phaseSettings());
    }

    public function part1(): string
    {
        return $this
            ->combinations
            ->map(fn(array $phase) => $this->runPhase($phase))
            ->max();
    }

    public function part2(): string
    {
        return
            $this
                ->combinations
                ->map(fn($x) => array_map(fn($y) => $y + 5, $x))
                ->map(fn(array $phase) => $this->runPhaseRecursive($phase))
                ->max();
    }

    private function phaseSettings(): \Generator
    {
        for ($a = 0; $a <= 4; $a++) {
            for ($b = 0; $b <= 4; $b++) {
                if ($b == $a) {
                    continue;
                }
                for ($c = 0; $c <= 4; $c++) {
                    if ($c == $a || $c == $b) {
                        continue;
                    }
                    for ($d = 0; $d <= 4; $d++) {
                        if ($d == $a || $d == $b || $d == $c) {
                            continue;
                        }
                        for ($e = 0; $e <= 4; $e++) {
                            if ($e == $a || $e == $b || $e == $c || $e == $d) {
                                continue;
                            }
                            yield [$a, $b, $c, $d, $e];
                        }
                    }
                }
            }
        }
    }

    private function runPhase(array $phases): int
    {
        $inputSignal = 0;
        foreach ($phases as $phase) {
            $this->intCode->reset();
            $this->intCode->inputList = [$phase, $inputSignal];
            $this->intCode->run();
            $inputSignal = $this->intCode->output;
        }
        return $inputSignal;
    }

    private function runPhaseRecursive(array $phase)
    {
        $this->intCode->reset();
        $computers = collect($phase)
            ->map(function ($p) {
                $clone = clone $this->intCode;
                $clone->inputList = [$p];
                return $clone;
            })
            ->tap(function (Collection $computers) {
                $computers->each(function (IntCode $computer, $index) use ($computers) {
                    $index += 1;
                    if ($index === $computers->count()) {
                        $index = 0;
                    }
                    $computer->outputCallable = function ($output) use ($computers, $index) {
                        $computers[$index]->inputList[] = $output;
                        $computers[$index]->run();
                    };
                });
            });

        $computers[0]->inputList[] = 0;
        $computers[0]->run();

        return $computers[4]->output;
    }
}

File: 2019/julia/src/day7.jl
include("Aoc.jl")
using .Aoc
include("Intcode.jl")
using .Intcode


function calculate_combinations()
    result = Set{Vector{Int64}}()
    for a in 0:4, b in 0:4, c in 0:4, d in 0:4, e in 0:4
        settings = [a,b,c,d,e]
        allunique(settings) || continue
        push!(result, settings)
    end
    result
end

combinations = calculate_combinations()

intcode = Aoc.intcode(7)

function run_combination(combination)
    inputSignal = 0
    for phase in combination
        c = Computer(copy(intcode))
        c.input = [phase, inputSignal]
        Intcode.run!(c)
        inputSignal = c.output
    end
    inputSignal
end

println("Part 1: ", maximum(run_combination.(combinations)))

part2_combinations = [c.+=5 for c in combinations]

function run_combination_recursive(combination)
    computers = Array{Computer,1}()
    for phase in combination
        computer = Computer(copy(intcode))
        push!(computer.input, phase)
        push!(computers, computer)
    end
    
    for (index, computer) = enumerate(computers)
        next_index = index + 1
        next_index > size(computers)[1] && (next_index = 1; true)
        computer.output_callable = (output) -> (push!(computers[next_index].input, output); Intcode.run!(computers[next_index]))
    end

    push!(computers[1].input, 0)
    Intcode.run!(computers[1])

    computers[5].output
end

println("Part 2: ", maximum(run_combination_recursive.(combinations)))